content/advent-2018/go-arrow.md

```diff
@@ -0,0 +1,531 @@
+Gophers might be surprised by this kind of low-level memory management, but this is needed to allow for `mmap`- or GPU-backed arrays.
+
+Another interesting piece of this interface is the `Data()` method that returns a value of type `*array.Data`:
+
+```go
+package array
+
+// A type which represents the memory and metadata for an Arrow array.
+type Data struct {
+	refCount int64
+	dtype    arrow.DataType
+	nulls    int
+	offset   int
+	length   int
+	buffers  []*memory.Buffer
+	kids     []*Data
+}
+```
+
+The `array.Data` type is where all the semantics of an Arrow arrays are encoded, as per the [Arrow specifications](https://arrow.apache.org/docs/memory_layout.html).
+
+The last piece to understand the memory layout of a Go Arrow array is `memory.Buffer`:
+
+```go
+package memory
+
+type Buffer struct {
+	refCount int64
+	buf      []byte
+	length   int
+	mutable  bool
+	mem      Allocator
+}
+```
+
+In a nutshell, Arrow arrays -- and the `array.Interface` -- can be seen as the cousins of the [buffer protocol](https://www.python.org/dev/peps/pep-3118/) from Python which has been instrumental in (pythonic) science.
+
+Let us come back to the `int32` array.
+It is implemented like so in Go:
+
+```go
+package array
+
+type Int32 struct {
+	array
+	values []int32
+}
+
+func NewInt32Data(data *Data) *Int32 { ... }
+
+type array struct {
+	refCount        int64
+	data            *Data
+	nullBitmapBytes []byte
+}
+```
+
+Creating a new array value holding `int32`s is done with `array.NewInt32Data(data)` where `data` holds a carefully crafted `array.Data` value with the needed memory buffers and sub-buffers (for the `null` bitmap.)
+As this can be a bit unwiedly, a helper type is provided to create Arrow arrays from scratch:
+
+```go
+mem := memory.NewGoAllocator()
+bld := array.NewInt32Builder(mem)
+defer bld.Release()
+
+// create an array with 4 values, no null
+bld.AppendValues([]int32{1, 2, 3, 4}, nil)
+
+arr1 := bld.NewInt32Array() // materialize the array
+defer arr1.Release()        // make sure we release memory, eventually.
+
+// arr1 = [1 2 3 4]
+fmt.Printf("arr1 = %v\n", arr1)
+
+// create an array with 5 values, 1 null
+bld.AppendValues(
+	[]int32{1, 2, 3, 4, 5},
+	[]bool{true, true, true, false, true},
+)
+
+arr2 := bld.NewInt32Array()
+defer arr2.Release()
+
+// arr2 = [1 2 3 (null) 5]
+fmt.Printf("arr2 = %v\n", arr2)
+```
+
+Arrow arrays can also be sub-sliced:
+
+```go
+sli := array.NewSlice(arr2, 1, 5).(*array.Int32)
+defer sli.Release()
+
+// slice = [2 3 (null) 5]
+fmt.Printf("slice = %v\n", sli)
+```
+
+Similar code can be written for arrays of `float64`, `float32`, unsigned integers, etc...
+But not everything can be mapped or simply expressed in terms of these basic primitives.
+What if we need a more detailed data type?
+That is expressed with `List`s or `Struct`s.
+Let us imagine we need to represent an array of "Person" type like:
+```go
+type Person struct {
+	Name string
+	Age  int32  // presumably, int8 should suffice.
```

Comment:
Add a field that contains a nested type (like a list of strings or integers), for emphasis?